include/oneapi/dpl/pstl/memory_ranges_impl.h

```diff
@@ -0,0 +1,202 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_MEMORY_RANGES_IMPL_H
+#define _ONEDPL_MEMORY_RANGES_IMPL_H
+
+#if _ONEDPL_CPP20_RANGES_PRESENT
+
+#    include <ranges>
+#    include <utility>
+#    include <cassert>
+#    include <functional>
+#    include <type_traits>
+
+#    include "execution_impl.h"
+#    include "glue_memory_impl.h"
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace __internal
+{
+namespace __ranges
+{
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_uninitialized_default_construct
+//---------------------------------------------------------------------------------------------------------------------
+
+template <typename _Tag, typename _ExecutionPolicy, typename _R>
+std::ranges::borrowed_iterator_t<_R>
+__pattern_uninitialized_default_construct(_Tag __tag, _ExecutionPolicy&& __exec, _R&& __r)
+{
+    const auto __first = std::ranges::begin(__r);
+    const auto __last = __first + std::ranges::size(__r);
+
+    oneapi::dpl::uninitialized_default_construct(std::forward<_ExecutionPolicy>(__exec), __first, __last);
```

Comment:
```suggestion
    oneapi::dpl::uninitialized_default_construct(std::forward<_ExecutionPolicy>(__exec), std::ranges::begin(__r), std::ranges::end(__r));
```

Reply:
Would you like to extract this evaluation into some separate function?

Reply:
#2383